### DISCUSSION TITLE
How to invoke Add/RemoveUser api in v5.0.3

### DISCUSSION THREAD

**Discussion by sshwy** [v2fly/v2ray-core#1756] [Q&A]
在尝试使用 grpc 调用 v2ray api，但是总是出现 `rpc error: code = Unknown desc = app/proxyman/command: unknown operation > Serial: Unknown type: types.v2fly.org/v2ray.core.app.proxyman.command.AddUserOperation` 的错误。

详细情况如下：拉取仓库的最新提交，本地构建 v2ray 可执行文件并配置好 api 参数。然后直接魔改 `testing/scenarios/command_test.go` [L277](https://github.com/v2fly/v2ray-core/blob/d7ad10ff14c82a6ee6f76a85834fe10b99706a27/testing/scenarios/command_test.go#L277) 把 cmdPort 改成 api 的端口，然后执行 `go test -v`。

想知道是我哪里没有做对，或者说有没有其他解决方案 >_<

P. S. 还要把 AddUserOperation 里的 [Tag](https://github.com/v2fly/v2ray-core/blob/d7ad10ff14c82a6ee6f76a85834fe10b99706a27/testing/scenarios/command_test.go#L283) （"v"）写到 v2ray 的某个 inbound 配置里

v2ray 配置：

```json
{
  "log": {
    "access": "/var/log/v2ray/access.log",
      "error": "/var/log/v2ray/error.log",
      "loglevel": "warning"
  },
    "stats": {},
    "api": {
      "tag": "api",
      "services": [
        "HandlerService",
      "LoggerService",
      "StatsService"
      ]
    },
  "routing": {
    "domainStrategy": "IPOnDemand",
      "rules": [
      {
        "inboundTag": [
          "api"
        ],
        "outboundTag": "api",
        "type": "field"
      }
    ]
  },
    "policy": {
      "levels": {
        "0": {
          "statsUserUplink": true,
          "statsUserDownlink": true
        }
      },
      "system": {
        "statsInboundUplink": true,
        "statsInboundDownlink": true,
        "statsOutboundUplink": true,
        "statsOutboundDownlink": true
      }
    },
    "inbounds": [
    {
      "listen": "127.0.0.1",
      "port": 10085,
      "protocol": "dokodemo-door",
      "settings": {
        "address": "127.0.0.1"
      },
      "tag": "api"
    },
    ...
    ],
    "outbounds": [
    {
      "protocol": "freedom",
      "settings": {}
    },
    {
      "protocol": "blackhole",
      "settings": {},
      "tag": "blocked"
    },
    {
      "protocol": "freedom",
      "settings": {},
      "tag": "direct"
    }
    ]
}

```

日志如下（只截了 TestCommanderAddRemoveUser）：

```
=== RUN   TestCommanderAddRemoveUser
Generated binary path: /tmp/v2ray2920307491/v2ray.test
Building V2Ray into path (/tmp/v2ray2920307491/v2ray.test)
V2Ray 5.0.3 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.18.1 linux/amd64)
A unified platform for anti-censorship.
2022/04/30 11:13:07 Using config from STDIN
2022/04/30 11:13:07 [Debug] app/proxyman/inbound: creating stream worker on 127.0.0.1:36103
2022/04/30 11:13:07 [Debug] app/proxyman/inbound: creating stream worker on 127.0.0.1:41139
V2Ray 5.0.3 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.18.1 linux/amd64)
A unified platform for anti-censorship.
2022/04/30 11:13:07 Using config from STDIN
2022/04/30 11:13:07 [Info] transport/internet/tcp: listening TCP on 127.0.0.1:36103
2022/04/30 11:13:07 [Info] transport/internet/tcp: listening TCP on 127.0.0.1:41139
2022/04/30 11:13:07 [Warning] V2Ray 5.0.3 started
2022/04/30 11:13:07 [Debug] app/proxyman/inbound: creating stream worker on 127.0.0.1:43321
2022/04/30 11:13:07 [Info] transport/internet/tcp: listening TCP on 127.0.0.1:43321
2022/04/30 11:13:07 [Warning] V2Ray 5.0.3 started
2022/04/30 11:13:09 [Debug] [1270167083] proxy/dokodemo: processing connection from: 127.0.0.1:36230
2022/04/30 11:13:09 [Info] [1270167083] proxy/dokodemo: received request for 127.0.0.1:36230
2022/04/30 11:13:09 [Warning] [1270167083] app/dispatcher: default route for tcp:127.0.0.1:35333
2022/04/30 11:13:09 127.0.0.1:36230 accepted tcp:127.0.0.1:35333
2022/04/30 11:13:09 [Info] [1270167083] transport/internet/tcp: dialing TCP to tcp:127.0.0.1:36103
2022/04/30 11:13:09 [Info] [1270167083] proxy/vmess/outbound: tunneling request to tcp:127.0.0.1:35333 via 127.0.0.1:36103
2022/04/30 11:13:09 127.0.0.1:50320 rejected  common/drain: common/drain: drained connection > proxy/vmess/encoding: invalid user > proxy/vmess: Not Found
2022/04/30 11:13:09 [Info] [1867379458] app/proxyman/inbound: connection ends > proxy/vmess/inbound: invalid request from 127.0.0.1:50320 > common/drain: common/drain: drained connection > proxy/vmess/encoding: invalid user > proxy/vmess: Not Found
2022/04/30 11:13:09 [Info] [1270167083] app/proxyman/outbound: failed to process outbound traffic > proxy/vmess/outbound: connection ends > context canceled
2022/04/30 11:13:09 [Info] [1270167083] app/proxyman/inbound: connection ends > proxy/dokodemo: connection ends > context canceled
testConn finishes: 2 ms	 EOF 	Alloc = 1.35MB 	TotalAlloc = 1.35MB 	Sys = 9.95MB 	NumGC = 0
2022/04/30 11:13:09 [Info] Closing all servers.
Failed accept TCP connection: accept tcp 127.0.0.1:35333: use of closed network connection
2022/04/30 11:13:09 [Info] All server closed.
--- FAIL: TestCommanderAddRemoveUser (5.33s)
panic: rpc error: code = Unknown desc = app/proxyman/command: unknown operation > Serial: Unknown type: types.v2fly.org/v2ray.core.app.proxyman.command.AddUserOperation [recovered]
	panic: rpc error: code = Unknown desc = app/proxyman/command: unknown operation > Serial: Unknown type: types.v2fly.org/v2ray.core.app.proxyman.command.AddUserOperation

goroutine 4 [running]:
testing.tRunner.func1.2({0xa9f340, 0xc00009c050})
	/usr/lib/go-1.18/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/usr/lib/go-1.18/src/testing/testing.go:1392 +0x39f
panic({0xa9f340, 0xc00009c050})
	/usr/lib/go-1.18/src/runtime/panic.go:838 +0x207
github.com/v2fly/v2ray-core/v5/common.Must(...)
	/root/v2ray-core/common/common.go:27
command-line-arguments.TestCommanderAddRemoveUser(0xc000217040)
	/root/v2ray-core/testing/scenarios/command_test.go:197 +0x18ed
testing.tRunner(0xc000217040, 0xb9eec8)
	/usr/lib/go-1.18/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/lib/go-1.18/src/testing/testing.go:1486 +0x35f
FAIL	command-line-arguments	5.342s
FAIL

```

**Comment by rayguo17** [comment]
Testing 的逻辑是会自己创建一个server和client，然后在自己创建的server和client之间进行互动不是嘛。这样的话，不会和本机运行的v2ray进行交互吧

**Comment by MJamshidnejad** [comment]
> 在尝试使用 grpc 调用 v2ray api，但是总是出现 `rpc error: code = Unknown desc = app/proxyman/command: unknown operation > Serial: Unknown type: types.v2fly.org/v2ray.core.app.proxyman.command.AddUserOperation` 的错误。
> 
> 详细情况如下：拉取仓库的最新提交，本地构建 v2ray 可执行文件并配置好 api 参数。然后直接魔改 `testing/scenarios/command_test.go` [L277](https://github.com/v2fly/v2ray-core/blob/d7ad10ff14c82a6ee6f76a85834fe10b99706a27/testing/scenarios/command_test.go#L277) 把 cmdPort 改成 api 的端口，然后执行 `go test -v`。
> 
> 想知道是我哪里没有做对，或者说有没有其他解决方案 >_<
> 
> P. S. 还要把 AddUserOperation 里的 [Tag](https://github.com/v2fly/v2ray-core/blob/d7ad10ff14c82a6ee6f76a85834fe10b99706a27/testing/scenarios/command_test.go#L283) （"v"）写到 v2ray 的某个 inbound 配置里
> 
> v2ray 配置：
> 
> ```json
> {
>   "log": {
>     "access": "/var/log/v2ray/access.log",
>       "error": "/var/log/v2ray/error.log",
>       "loglevel": "warning"
>   },
>     "stats": {},
>     "api": {
>       "tag": "api",
>       "services": [
>         "HandlerService",
>       "LoggerService",
>       "StatsService"
>       ]
>     },
>   "routing": {
>     "domainStrategy": "IPOnDemand",
>       "rules": [
>       {
>         "inboundTag": [
>           "api"
>         ],
>         "outboundTag": "api",
>         "type": "field"
>       }
>     ]
>   },
>     "policy": {
>       "levels": {
>         "0": {
>           "statsUserUplink": true,
>           "statsUserDownlink": true
>         }
>       },
>       "system": {
>         "statsInboundUplink": true,
>         "statsInboundDownlink": true,
>         "statsOutboundUplink": true,
>         "statsOutboundDownlink": true
>       }
>     },
>     "inbounds": [
>     {
>       "listen": "127.0.0.1",
>       "port": 10085,
>       "protocol": "dokodemo-door",
>       "settings": {
>         "address": "127.0.0.1"
>       },
>       "tag": "api"
>     },
>     ...
>     ],
>     "outbounds": [
>     {
>       "protocol": "freedom",
>       "settings": {}
>     },
>     {
>       "protocol": "blackhole",
>       "settings": {},
>       "tag": "blocked"
>     },
>     {
>       "protocol": "freedom",
>       "settings": {},
>       "tag": "direct"
>     }
>     ]
> }
> ```
> 
> 日志如下（只截了 TestCommanderAddRemoveUser）：
> 
> ```
> === RUN   TestCommanderAddRemoveUser
> Generated binary path: /tmp/v2ray2920307491/v2ray.test
> Building V2Ray into path (/tmp/v2ray2920307491/v2ray.test)
> V2Ray 5.0.3 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.18.1 linux/amd64)
> A unified platform for anti-censorship.
> 2022/04/30 11:13:07 Using config from STDIN
> 2022/04/30 11:13:07 [Debug] app/proxyman/inbound: creating stream worker on 127.0.0.1:36103
> 2022/04/30 11:13:07 [Debug] app/proxyman/inbound: creating stream worker on 127.0.0.1:41139
> V2Ray 5.0.3 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.18.1 linux/amd64)
> A unified platform for anti-censorship.
> 2022/04/30 11:13:07 Using config from STDIN
> 2022/04/30 11:13:07 [Info] transport/internet/tcp: listening TCP on 127.0.0.1:36103
> 2022/04/30 11:13:07 [Info] transport/internet/tcp: listening TCP on 127.0.0.1:41139
> 2022/04/30 11:13:07 [Warning] V2Ray 5.0.3 started
> 2022/04/30 11:13:07 [Debug] app/proxyman/inbound: creating stream worker on 127.0.0.1:43321
> 2022/04/30 11:13:07 [Info] transport/internet/tcp: listening TCP on 127.0.0.1:43321
> 2022/04/30 11:13:07 [Warning] V2Ray 5.0.3 started
> 2022/04/30 11:13:09 [Debug] [1270167083] proxy/dokodemo: processing connection from: 127.0.0.1:36230
> 2022/04/30 11:13:09 [Info] [1270167083] proxy/dokodemo: received request for 127.0.0.1:36230
> 2022/04/30 11:13:09 [Warning] [1270167083] app/dispatcher: default route for tcp:127.0.0.1:35333
> 2022/04/30 11:13:09 127.0.0.1:36230 accepted tcp:127.0.0.1:35333
> 2022/04/30 11:13:09 [Info] [1270167083] transport/internet/tcp: dialing TCP to tcp:127.0.0.1:36103
> 2022/04/30 11:13:09 [Info] [1270167083] proxy/vmess/outbound: tunneling request to tcp:127.0.0.1:35333 via 127.0.0.1:36103
> 2022/04/30 11:13:09 127.0.0.1:50320 rejected  common/drain: common/drain: drained connection > proxy/vmess/encoding: invalid user > proxy/vmess: Not Found
> 2022/04/30 11:13:09 [Info] [1867379458] app/proxyman/inbound: connection ends > proxy/vmess/inbound: invalid request from 127.0.0.1:50320 > common/drain: common/drain: drained connection > proxy/vmess/encoding: invalid user > proxy/vmess: Not Found
> 2022/04/30 11:13:09 [Info] [1270167083] app/proxyman/outbound: failed to process outbound traffic > proxy/vmess/outbound: connection ends > context canceled
> 2022/04/30 11:13:09 [Info] [1270167083] app/proxyman/inbound: connection ends > proxy/dokodemo: connection ends > context canceled
> testConn finishes: 2 ms	 EOF 	Alloc = 1.35MB 	TotalAlloc = 1.35MB 	Sys = 9.95MB 	NumGC = 0
> 2022/04/30 11:13:09 [Info] Closing all servers.
> Failed accept TCP connection: accept tcp 127.0.0.1:35333: use of closed network connection
> 2022/04/30 11:13:09 [Info] All server closed.
> --- FAIL: TestCommanderAddRemoveUser (5.33s)
> panic: rpc error: code = Unknown desc = app/proxyman/command: unknown operation > Serial: Unknown type: types.v2fly.org/v2ray.core.app.proxyman.command.AddUserOperation [recovered]
> 	panic: rpc error: code = Unknown desc = app/proxyman/command: unknown operation > Serial: Unknown type: types.v2fly.org/v2ray.core.app.proxyman.command.AddUserOperation
> 
> goroutine 4 [running]:
> testing.tRunner.func1.2({0xa9f340, 0xc00009c050})
> 	/usr/lib/go-1.18/src/testing/testing.go:1389 +0x24e
> testing.tRunner.func1()
> 	/usr/lib/go-1.18/src/testing/testing.go:1392 +0x39f
> panic({0xa9f340, 0xc00009c050})
> 	/usr/lib/go-1.18/src/runtime/panic.go:838 +0x207
> github.com/v2fly/v2ray-core/v5/common.Must(...)
> 	/root/v2ray-core/common/common.go:27
> command-line-arguments.TestCommanderAddRemoveUser(0xc000217040)
> 	/root/v2ray-core/testing/scenarios/command_test.go:197 +0x18ed
> testing.tRunner(0xc000217040, 0xb9eec8)
> 	/usr/lib/go-1.18/src/testing/testing.go:1439 +0x102
> created by testing.(*T).Run
> 	/usr/lib/go-1.18/src/testing/testing.go:1486 +0x35f
> FAIL	command-line-arguments	5.342s
> FAIL
> ```

Dear @sshwy , did you do that in previous versions? like 4.45.2? 
I mean adding users to an inbound.
